servers/src/grin/dandelion_monitor.rs

```diff
@@ -44,211 +46,153 @@ pub fn monitor_transactions(
-				// We will process these in the next run (waiting patience secs).
-				if process_fresh_entries(dandelion_config.clone(), tx_pool.clone()).is_err() {
-					error!("dand_mon: Problem processing fresh pool entries.");
+				// Handle the tx above *before* we transition to next epoch.
+				// This gives us an opportunity to do the final "fluff" before we start
+				// stemming on the subsequent epoch.
+				if adapter.is_expired() {
+					adapter.next_epoch();
 				}
 
-				// Step 4: now find all expired entries based on embargo timer.
-				if process_expired_entries(dandelion_config.clone(), tx_pool.clone()).is_err() {
-					error!("dand_mon: Problem processing fresh pool entries.");
-				}
+				// Monitor loops every 10s.
+				thread::sleep(Duration::from_secs(10));
 			}
 		});
 }
 
-fn process_stem_phase(
-	tx_pool: Arc<RwLock<TransactionPool>>,
-	verifier_cache: Arc<RwLock<dyn VerifierCache>>,
+// Query the pool for transactions older than the cutoff.
+// Used for both periodic fluffing and handling expired embargo timer.
+fn select_txs_cutoff(pool: &Pool, cutoff_secs: u64) -> Vec<PoolEntry> {
+	let cutoff = Utc::now().timestamp() - cutoff_secs as i64;
+	pool.entries
+		.iter()
+		.filter(|x| x.tx_at.timestamp() < cutoff)
+		.cloned()
+		.collect()
+}
+
+fn process_fluff_phase(
+	dandelion_config: &DandelionConfig,
+	tx_pool: &Arc<RwLock<TransactionPool>>,
+	adapter: &Arc<DandelionAdapter>,
+	verifier_cache: &Arc<RwLock<dyn VerifierCache>>,
 ) -> Result<(), PoolError> {
+	// Take a write lock on the txpool for the duration of this processing.
 	let mut tx_pool = tx_pool.write();
 
-	let header = tx_pool.chain_head()?;
-
-	let stem_txs = tx_pool
-		.stempool
-		.get_transactions_in_state(PoolEntryState::ToStem);
+	// if we are expired the look for all txs.
+	// if not expired then -
+	// if multiple look for all txs
+	// if single then wait 30s
 
-	if stem_txs.is_empty() {
+	let all_entries = select_txs_cutoff(&tx_pool.stempool, 0);
+	if all_entries.is_empty() {
 		return Ok(());
 	}
 
-	// Get the aggregate tx representing the entire txpool.
-	let txpool_tx = tx_pool.txpool.all_transactions_aggregate()?;
-
-	let stem_txs = tx_pool
-		.stempool
-		.select_valid_transactions(stem_txs, txpool_tx, &header)?;
-	tx_pool
-		.stempool
-		.transition_to_state(&stem_txs, PoolEntryState::Stemmed);
-
-	if stem_txs.len() > 0 {
-		debug!("dand_mon: Found {} txs for stemming.", stem_txs.len());
-
-		let agg_tx = transaction::aggregate(stem_txs)?;
-		agg_tx.validate(
-			transaction::Weighting::AsTransaction,
-			verifier_cache.clone(),
-		)?;
-
-		let res = tx_pool.adapter.stem_tx_accepted(&agg_tx);
-		if res.is_err() {
-			debug!("dand_mon: Unable to propagate stem tx. No relay, fluffing instead.");
+	let cutoff_secs = dandelion_config
+		.aggregation_secs
+		.expect("aggregation secs config missing");
+	let cutoff_entries = select_txs_cutoff(&tx_pool.stempool, cutoff_secs);
+
+	if adapter.is_expired() {
```

Comment:
Did you mean to keep those conditions around the comments? Seem that code doc shouldn't rely on actual code :-)

Reply:
Hmm - tried to make it clear what was going on here. Now I'm looking at fresh I see what you are saying. I'll rewrite this and clean it up.